test_workspace/BinNat.bosatsu

```diff
@@ -97,11 +142,51 @@ def times2(b: BinNat) -> BinNat:
       #2(2(n + 1)) = 2((2n + 1) + 1)
       Even(Odd(n))
 
+def sub_BinNat(left: BinNat, right: BinNat) -> BinNat:
+  # invariant: left > right
+  # we always return >= 1
+  def loop(left, right):
+    recur left:
+      case Odd(left) as odd:
+        match right:
+          case Zero: odd
+          case Odd(right):
+            # (2n + 1) - (2m + 1) = 2(n - m)
+            times2(loop(left, right))
+          case Even(right):
+            # (2n + 1) - (2m + 2) = 2(n - m) - 1
+            # note if (2n + 1) > (2m + 2), then n > m
+            times2(loop(left, right)).prev()
+      case Even(left) as even:
+        match right:
+          case Zero: even
+          case Odd(right):
+            # (2n + 2) - (2m + 1)
+            # if n >= m: 2(n - m) + 1
+            # if (2n + 2) > (2m + 1), then n > m
```

Comment:
this is wrong... if (2n + 2) > (2m + 1) then n >= m. So, that means the invariant must actually be left >= right. So that means we only know that (2n + 2) >= (2m + 1), but that's okay, because we know an odd number can't equal an even number, so we recover n >= m.